samples/nrf9160/modem_trace_flash/src/trace_flash.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/storage/stream_flash.h>
+#include <zephyr/storage/flash_map.h>
+#include <zephyr/logging/log.h>
+#include <modem/trace_backend.h>
+
+#include "trace_flash.h"
+
+LOG_MODULE_REGISTER(modem_trace_flash_backend, CONFIG_MODEM_TRACE_FLASH_BACKEND_LOG_LEVEL);
+
+#define EXT_FLASH_DEVICE DEVICE_DT_GET(DT_ALIAS(ext_flash))
+#define TRACE_OFFSET FLASH_AREA_OFFSET(modem_trace)
+#define TRACE_SIZE FLASH_AREA_SIZE(modem_trace)
+#define BUF_LEN 1024
+
+static const struct flash_area *modem_trace_area;
+static const struct device *flash_dev;
+static struct stream_flash_ctx stream;
+static uint8_t stream_buf[BUF_LEN];
+
+// Persist trace size in .noinit RAM to survive warm boot.
+static __noinit uint32_t magic;
+static __noinit uint32_t traces_size;
+
+static int64_t tot_write_time;
+static trace_backend_processed_cb trace_processed_callback;
+
+static int internal_flash_write(const void *data, size_t len, bool flush)
+{
+	int64_t uptime_ref = k_uptime_get();
+	int64_t write_time;
+
+	int err = stream_flash_buffered_write(&stream, data, len, flush);
+
+	if (err != 0) {
+		LOG_ERR("stream_flash_buffered_write error %d", err);
+		return err;
+	}
+
+	write_time = k_uptime_delta(&uptime_ref);
+	tot_write_time += write_time;
+
+	traces_size += len;
+
+	return (int)len;
+}
+
+static const uint32_t magic_val = 0x152ac523;
+int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
+{
+	int err;
+
+	LOG_DBG("trace_backend_init\n");
+	if (trace_processed_cb == NULL) {
+		return -EFAULT;
+	}
+
+	trace_processed_callback = trace_processed_cb;
+
+	if (magic != magic_val) {
+		LOG_DBG("Initializing .noinit\n");
+		magic = magic_val;
+		traces_size = 0;
+	}
+
+	traces_size = 0;
+
+	err = flash_area_open(FLASH_AREA_ID(modem_trace), &modem_trace_area);
+	if (err) {
+		LOG_ERR("flash_area_open error:  %d", err);
+		return -ENODEV;
+	}
+
+	err = flash_area_has_driver(modem_trace_area);
+	if (err != 1) {
+		LOG_ERR("flash_area_has_driver error: %d\n", err);
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	err = stream_flash_init(&stream, flash_dev, stream_buf, BUF_LEN, TRACE_OFFSET, TRACE_SIZE,
+		NULL);
+	if (err) {
+		LOG_ERR("stream_flash_init error: %d", err);
+		return err;
+	}
+
+	// erase flash partition first to write without interruptions while tracing
```

Comment:
You can remove this comment now. We dont erase the partition to avoid interruptions do we? We rather prepare the flash.

Reply:
We do erase first, but I can improve the comment.